service/dap/server.go

```diff
@@ -1481,6 +1450,81 @@ func (s *Server) onEvaluateRequest(request *dap.EvaluateRequest) {
 	s.send(response)
 }
 
+func (s *Server) doCall(goid, frame int, expr string) ([]*proc.Variable, error) {
+	// This call might be evaluated in the context of the frame that is not topmost
+	// if the editor is set to view the variables for one of the parent frames.
+	// If the call expression refers to any of these variables, unlike regular
+	// expressions, it will evaluate them in the context of the topmost frame,
+	// and the user will get an unexpected result or an unexpected symbol error.
+	// We prevent this but disallowing any frames other than topmost.
+	if frame > 0 {
+		return nil, fmt.Errorf("call is only supported with topmost stack frame")
+	}
+	stateBeforeCall, err := s.debugger.State( /*nowait*/ true)
+	if err != nil {
+		return nil, err
+	}
+	state, err := s.debugger.Command(&api.DebuggerCommand{
+		Name:                 api.Call,
+		ReturnInfoLoadConfig: api.LoadConfigFromProc(&DefaultLoadConfig),
+		Expr:                 expr,
+		UnsafeCall:           false,
+		GoroutineID:          goid,
+	}, nil)
+	if _, isexited := err.(proc.ErrProcessExited); isexited || err == nil && state.Exited {
+		return nil, errTerminated
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// Call commands used for variable assignments do not return a value.
+	// In go '=' is not an operator. Check if go/parser complains.
+	// If the above Call command passed but the expression is not a valid
+	// go expression, we just handled an assignment, and there is no
+	// return value for assignment.
+	if _, err := parser.ParseExpr(expr); err != nil {
+		s.resetHandlesForStop()
+		s.sendStoppedEvent(state)
+		return nil, nil
+	}
+
+	// After the call is done, the goroutine where we injected the call should
+	// return to the original stopped line with return values. However,
+	// it is not guaranteed to be selected due to the possibility of the
+	// of simultaenous breakpoints. Therefore, we check all threads.
+	var retVars []*proc.Variable
+	for _, t := range state.Threads {
+		if t.GoroutineID == stateBeforeCall.SelectedGoroutine.ID &&
+			t.Line == stateBeforeCall.SelectedGoroutine.CurrentLoc.Line && t.CallReturn {
+			// The call completed. Get the return values.
+			retVars, err = s.debugger.FindThreadReturnValues(t.ID, DefaultLoadConfig)
+			if err != nil {
+				return nil, err
+			}
+			break
+		}
+	}
+	if retVars == nil {
+		// The call got interrupted by a stop (e.g. breakpoint in injected
+		// function call or in another goroutine).
+		s.resetHandlesForStop()
```

Comment:
I think this should be done unconditionally, every time continue is called. Variable objects in particular aren't guaranteed to survive a continue, especially now that SetVariable is supported: invalid Variables could lead to memory corruption.

Reply:
Good point. Left a TODO. For other continues (doCommand is already doing the job I believe). For this call command, this is a pre-existing bug and I hope to address it separately.